src/lua/protobuf_wireformat.lua

```diff
@@ -0,0 +1,147 @@
+    if type(value) == 'number' then
+        local ctype = value < 0 and int64_t or uint64_t
+        value = ffi.cast(ctype, value)
+    end
+    repeat
+        local payload = bit.bor(bit.band(value, 0x7f), 0x80)
+        value = bit.rshift(value, 7)
+        buf[size] = payload
+        size = size + 1
+    until value == 0
+    buf[size-1] = bit.band(buf[size-1], 0x7f)
+    return ffi.string(buf, size)
+end
+
+-- Encode a tag byte from the given field_id and the given Protocol Buffers
+-- wire type. This is the first byte of Tag-Length-Value encoding.
+local function encode_tag(field_id, wire_type)
+    assert(wire_type >= 0 and wire_type <= 5)
+    return encode_varint(bit.bor(bit.lshift(field_id, 3), wire_type))
+end
+
+-- }}} Helpers
+
+-- {{{ API functions
+
+-- Encode an integral value as VARINT using two complement encoding.
+-- Input value types: number (integral), cdata<int64_t>, cdata<uint64_t>,
+-- boolean.
+-- Used for Protocol Buffers types: int32, int64, uint32, uint64, bool, enum.
+local function encode_int(field_id, value)
+    if type(value) == 'boolean' then
+        value = value and 1 or 0
+    end
+    return encode_tag(field_id, WIRE_TYPE_VARINT) .. encode_varint(value)
+end
+
+-- Encode an integral value as VARINT using the "ZigZag" encoding.
+-- Input value types: number (integral), cdata<int64_t>, cdata<uint64_t>.
+-- Used for Protocol Buffers types: sint32, sint64.
+local function encode_sint(field_id, value)
+    if value >= 0 then
+        return encode_int(field_id, 2 * value)
+    else
+        return encode_int(field_id, 2 * (-value) - 1)
+    end
+end
+
+-- Encode an integral value as I32.
+-- Input value types: number (integral), cdata<int64_t>, cdata<uint64_t>.
+-- Used for Protocol Buffers types: fixed32, sfixed32.
+local function encode_fixed32(field_id, value)
+    local buf = ffi.new('char[?]', 5)
+    local size = 0
+    value = ffi.new('uint32_t', value)
+    repeat
+        local payload = bit.band(value, 0xff)
+        value = bit.rshift(value, 8)
+        buf[size] = payload
+        size = size + 1
+    until size == 4
+    return encode_tag(field_id, WIRE_TYPE_I32) .. ffi.string(buf, size)
+end
+
+-- Encode a floating point value as I32.
+-- Input value type: number.
+-- Used for Protocol Buffers type: float.
+local function encode_float(field_id, value)
+    return encode_tag(field_id, WIRE_TYPE_I32) .. as_float(value)
+end
+
+-- Encode an integral value as I64.
+-- Input value types: number (integral), cdata<int64_t>, cdata<uint64_t>.
+-- Used for Protocol Buffers types: fixed64, sfixed64.
+local function encode_fixed64(field_id, value)
+    local buf = ffi.new('char[?]', 9)
+    local size = 0
+    value = ffi.new('uint64_t', value)
+    repeat
+        local payload = bit.band(value, 0xff)
+        value = bit.rshift(value, 8)
+        buf[size] = payload
+        size = size + 1
+    until size == 8
+    return encode_tag(field_id, WIRE_TYPE_I64) .. ffi.string(buf, size)
+end
```

Comment:
We don't need neither loop, nor bit operations here. We can use the same trick as in `as_float` and `as_double`.

```diff
diff --git a/src/lua/protobuf_wireformat.lua b/src/lua/protobuf_wireformat.lua
index 3139835f12..e10d79748b 100644
--- a/src/lua/protobuf_wireformat.lua
+++ b/src/lua/protobuf_wireformat.lua
@@ -10,6 +10,48 @@ local WIRE_TYPE_I32 = 5
 
 -- {{{ Helpers
 
+-- 32-bit two's complement representation of the given integral
+-- number.
+local function as_int32(value)
+    -- Determine appropriate storage for the given value: signed
+    -- or unsigned 32-bit type.
+    local ctype
+    if type(value) == 'number' then
+        ctype = value < 0 and 'int32_t[1]' or 'uint32_t[1]'
+    elseif type(value) == 'cdata' and ffi.istype(int64_t, value) then
+        ctype = 'int32_t[1]'
+    elseif type(value) == 'cdata' and ffi.istype(uint64_t, value) then
+        ctype = 'uint32_t[1]'
+    else
+        assert(false)
+    end
+
+    local p = ffi.new(ctype)
+    p[0] = value
+    return ffi.string(ffi.cast('char *', p), 4)
+end
+
+-- 64-bit two's complement representation of the given integral
+-- number.
+local function as_int64(value)
+    -- Determine appropriate storage for the given value: signed
+    -- or unsigned 64-bit type.
+    local ctype
+    if type(value) == 'number' then
+        ctype = value < 0 and 'int64_t[1]' or 'uint64_t[1]'
+    elseif type(value) == 'cdata' and ffi.istype(int64_t, value) then
+        ctype = 'int64_t[1]'
+    elseif type(value) == 'cdata' and ffi.istype(uint64_t, value) then
+        ctype = 'uint64_t[1]'
+    else
+        assert(false)
+    end
+
+    local p = ffi.new(ctype)
+    p[0] = value
+    return ffi.string(ffi.cast('char *', p), 8)
+end
+
 -- 32-bit IEEE 754 representation of the given number.
 local function as_float(value)
     local p = ffi.new('float[1]')
@@ -81,16 +123,7 @@ end
 -- Input value types: number (integral), cdata<int64_t>, cdata<uint64_t>.
 -- Used for Protocol Buffers types: fixed32, sfixed32.
 local function encode_fixed32(field_id, value)
-    local buf = ffi.new('char[?]', 5)
-    local size = 0
-    value = ffi.new('uint32_t', value)
-    repeat
-        local payload = bit.band(value, 0xff)
-        value = bit.rshift(value, 8)
-        buf[size] = payload
-        size = size + 1
-    until size == 4
-    return encode_tag(field_id, WIRE_TYPE_I32) .. ffi.string(buf, size)
+    return encode_tag(field_id, WIRE_TYPE_I32) .. as_int32(value)
 end
 
 -- Encode a floating point value as I32.
@@ -104,16 +137,7 @@ end
 -- Input value types: number (integral), cdata<int64_t>, cdata<uint64_t>.
 -- Used for Protocol Buffers types: fixed64, sfixed64.
 local function encode_fixed64(field_id, value)
-    local buf = ffi.new('char[?]', 9)
-    local size = 0
-    value = ffi.new('uint64_t', value)
-    repeat
-        local payload = bit.band(value, 0xff)
-        value = bit.rshift(value, 8)
-        buf[size] = payload
-        size = size + 1
-    until size == 8
-    return encode_tag(field_id, WIRE_TYPE_I64) .. ffi.string(buf, size)
+    return encode_tag(field_id, WIRE_TYPE_I64) .. as_int64(value)
 end
 
 -- Encode a floating point value as I64.
```

Reply:
Implemented `as_int32` and `as_int64` functions for `encode_fixed32` and `encode_fixed64`